tests/integration.rs

```diff
@@ -0,0 +1,433 @@
+
+        let body = json!({
+            "token": token,
+            "content": key_auth,
+        })
+        .to_string();
+
+        let req = Request::builder()
+            .method(Method::POST)
+            .uri(format!("{}/add-http01", self.challenge_management_url()))
+            .header(CONTENT_TYPE, "application/json")
+            .body(Full::new(body.into()))?;
+
+        client.request(req).await?;
+
+        Ok(())
+    }
+
+    // Return a RootCertStore containing the issuer root for the Pebble CA.
+    //
+    // This is distinct from the management root CA, and is randomly generated each
+    // time that Pebble starts up. This is the issuer that signs the randomly generated
+    // intermediate certificate returned as part of ACME issued certificate chains.
+    async fn issuer_roots(&self) -> Result<RootCertStore, Box<dyn StdError>> {
+        let client = self.http_client();
+
+        let req = Request::builder()
+            .method(Method::GET)
+            .uri(format!("{}/roots/0", self.pebble_management_url()))
+            .header(CONTENT_TYPE, "application/json")
+            .body(Full::default())?;
+
+        let resp = client.request(req).await?;
+        if resp.status() != 200 {
+            return Err(format!("unexpected /roots/0 response status: {}", resp.status()).into());
+        }
+
+        let body = resp.collect().await?.aggregate();
+        let mut pem = String::new();
+        body.reader().read_to_string(&mut pem)?;
+        let root = CertificateDer::from_pem_slice(pem.as_bytes())?;
+
+        let mut roots = RootCertStore::empty();
+        roots.add_parsable_certificates(vec![root]);
+        assert_eq!(roots.len(), 1);
+        Ok(roots)
+    }
+
+    fn challenge_management_url(&self) -> &str {
+        "http://127.0.0.1:8055" // Default.
+    }
+
+    fn pebble_management_url(&self) -> String {
+        format!("https://{}", &self.config.pebble.management_listen_address)
+    }
+
+    fn directory_url(&self) -> String {
+        format!("https://{}/dir", &self.config.pebble.listen_address)
+    }
+}
+
+#[derive(Clone, Serialize)]
+struct Config {
+    pebble: PebbleConfig,
+}
+
+#[derive(Clone, Serialize)]
+struct PebbleConfig {
+    #[serde(rename = "listenAddress")]
+    listen_address: String,
+    #[serde(rename = "managementListenAddress")]
+    management_listen_address: String,
+    certificate: String,
+    #[serde(rename = "privateKey")]
+    private_key: String,
+    #[serde(rename = "httpPort")]
+    http_port: u16,
+    #[serde(rename = "tlsPort")]
+    tls_port: u16,
+    #[serde(rename = "ocspResponderURL")]
+    ocsp_responder_url: String,
+    #[serde(rename = "externalAccountBindingRequired")]
+    external_account_binding_required: bool,
+    #[serde(rename = "domainBlocklist")]
+    domain_blocklist: Vec<String>,
+    retry_after: RetryConfig,
+    profiles: HashMap<String, Profile>,
+}
+
+#[derive(Clone, Serialize)]
+struct RetryConfig {
+    authz: u32,
```

Comment:
Maybe add a docstring to specify the units, or use `Duration` and a `serde(serialize_with = "")`?

Reply:
Added docstrings.